framework/stats/hwcpipe_stats_provider.cpp

```diff
@@ -0,0 +1,218 @@
+	    {StatIndex::gpu_ext_read_bytes,    {hwcpipe::GpuCounter::ExternalMemoryReadBytes}},
+	    {StatIndex::gpu_ext_write_bytes,   {hwcpipe::GpuCounter::ExternalMemoryWriteBytes}},
+	    {StatIndex::gpu_tex_cycles,        {hwcpipe::GpuCounter::ShaderTextureCycles}}};
+	// clang-format on
+
+	hwcpipe::CpuCounterSet enabled_cpu_counters{};
+	hwcpipe::GpuCounterSet enabled_gpu_counters{};
+
+	for (const auto &stat : requested_stats)
+	{
+		auto res = hwcpipe_stats.find(stat);
+		if (res != hwcpipe_stats.end())
+		{
+			stat_data[stat] = hwcpipe_stats[stat];
+
+			switch (res->second.type)
+			{
+				case StatType::Cpu:
+					enabled_cpu_counters.insert(res->second.cpu_counter);
+					if (res->second.divisor_cpu_counter != hwcpipe::CpuCounter::MaxValue)
+						enabled_cpu_counters.insert(res->second.divisor_cpu_counter);
+					break;
+				case StatType::Gpu:
+					enabled_gpu_counters.insert(res->second.gpu_counter);
+					if (res->second.divisor_gpu_counter != hwcpipe::GpuCounter::MaxValue)
+						enabled_gpu_counters.insert(res->second.divisor_gpu_counter);
+					break;
+			}
+		}
+	}
+
+	hwcpipe = std::make_unique<hwcpipe::HWCPipe>(enabled_cpu_counters, enabled_gpu_counters);
+
+	// Now that we've made a hwcpipe with the counters we'd like, remove any that
+	// aren't actually supported
+	for (auto iter = stat_data.begin(); iter != stat_data.end();)
+	{
+		switch (iter->second.type)
+		{
+			case StatType::Cpu:
+			{
+				if (hwcpipe->cpu_profiler())
+				{
+					const auto &cpu_supp = hwcpipe->cpu_profiler()->supported_counters();
+					if (cpu_supp.find(iter->second.cpu_counter) == cpu_supp.end())
+						iter = stat_data.erase(iter);
+					else
+						++iter;
+				}
+				else
+					iter = stat_data.erase(iter);
+				break;
+			}
+			case StatType::Gpu:
+			{
+				if (hwcpipe->gpu_profiler())
+				{
+					const auto &gpu_supp = hwcpipe->gpu_profiler()->supported_counters();
+					if (gpu_supp.find(iter->second.gpu_counter) == gpu_supp.end())
+						iter = stat_data.erase(iter);
+					else
+						++iter;
+				}
+				else
+					iter = stat_data.erase(iter);
+				break;
+			}
+		}
+	}
+
+	// Remove any supported stats from the requested set.
+	// Subsequent providers will then only look for things that aren't already supported.
+	for (const auto &iter : stat_data)
+	{
+		requested_stats.erase(iter.first);
+	}
+
+	hwcpipe->run();
+}
+
+bool HWCPipeStatsProvider::is_available(StatIndex index) const
+{
+	return stat_data.find(index) != stat_data.end();
+}
+
+const StatGraphData &HWCPipeStatsProvider::get_graph_data(StatIndex index) const
+{
+	static StatGraphData vertex_compute_cycles{"Vertex Compute Cycles", "{:4.1f} M/s", float(1e-6)};
+
+	assert(is_available(index) && "HWCPipeStatsProvider::get_graph_data() called with invalid StatIndex");
+
+	// HWCPipe reports combined vertex/compute cycles (which is Arm specific)
+	// Ensure we report graph that with the correct name when asked for vertex cycles
```

Comment:
From what I can see, VulkanException is generally thrown when a Vulkan call or Vulkan creation fails. This assert is really checking that the framework isn't calling this without first checking for its availability. It doesn't feel in the same class as a VulkanException to me. Since it should never happen an assert, which only happens in debug builds, feels more correct.

Reply:
I've moved the assert though.

Reply:
Yes that makes sense, thanks!